cluster-autoscaler/processors/nodegroupset/balancing_processor.go

```diff
@@ -179,5 +193,66 @@ func (b *BalancingNodeGroupSetProcessor) BalanceScaleUpBetweenGroups(context *co
 	return result, nil
 }
 
+// averageNodegroupAllocatableResources collects the average capacity by nodegroup id,
+// so we can add new nodes to the one with the lowest capacity instead of the one with the lowest node count
+func (b *BalancingNodeGroupSetProcessor) averageNodegroupAllocatableResources(context *context.AutoscalingContext, groups []cloudprovider.NodeGroup, nodes []*apiv1.Node) map[string]int {
+	// collect info
+	averageNodegroupCapacities := map[string]int{}
+	nodegroupCounts := map[string]int{}
+	fallback := 0
+
+	if context.BalanceSimilarNodeGroupsBy == BalanceByCpu || context.BalanceSimilarNodeGroupsBy == BalanceByMemory {
+		for _, ng := range groups {
+			averageNodegroupCapacities[ng.Id()] = 0
+			nodegroupCounts[ng.Id()] = 0
+			ngNodes, err := ng.Nodes()
+			if err != nil {
+				klog.V(1).Infof("Failed to get nodes from ASG %v: %v", ng.Id(), err)
+				continue
+			}
+			for _, n := range nodes {
+				nId := n.Spec.ProviderID
+				for _, ngNode := range ngNodes {
+					if nId == ngNode.Id {
+						nodegroupCounts[ng.Id()]++
+						var resource *resource.Quantity
+						if context.BalanceSimilarNodeGroupsBy == "cpu" {
+							resource = n.Status.Allocatable.Cpu()
+						} else {
+							resource = n.Status.Allocatable.Memory()
+						}
+						averageNodegroupCapacities[ng.Id()] += int(resource.MilliValue())
+					}
+				}
+			}
+		}
+
+		// calculate average and collect fallback
+		for _, ng := range groups {
+			if nodegroupCounts[ng.Id()] == 0 {
+				klog.V(3).Infof("Failed to find nodes for ASG %v: %v", ng.Id())
+			} else {
+				averageNodegroupCapacities[ng.Id()] /= nodegroupCounts[ng.Id()]
+				fallback = averageNodegroupCapacities[ng.Id()]
```

Comment:
So fallback is the average capacity in a first (== random) node group? Shouldn't it be average across known node groups?

Reply:
yeah, good catch!